core/src/main/java/com/linecorp/armeria/client/endpoint/WeightRampingUpStrategy.java

```diff
@@ -0,0 +1,493 @@
+            if (endpointGroup instanceof DynamicEndpointGroup) {
+                ((DynamicEndpointGroup) endpointGroup).whenClosed().thenRunAsync(this::close, executor);
+            }
+        }
+
+        /**
+         * Removes the duplicate endpoints in the specified {@code newEndpoints} and returns a new map
+         * that contains unique endpoints.
+         * The value of the map is the {@link Endpoint} whose {@link Endpoint#weight()} is the summed weight of
+         * same {@link Endpoint}s.
+         */
+        private Map<Endpoint, Endpoint> removeDuplicateEndpoints(List<Endpoint> newEndpoints) {
+            final Map<Endpoint, Endpoint> newEndpointsMap = new HashMap<>(newEndpoints.size());
+
+            // The weight of the same endpoints are summed.
+            newEndpoints.forEach(
+                    newEndpoint -> newEndpointsMap.compute(newEndpoint, (key, v) -> {
+                        if (v == null) {
+                            return newEndpoint;
+                        }
+                        final int weightSum = newEndpoint.weight() + v.weight();
+                        return newEndpoint.withWeight(weightSum);
+                    }));
+            return newEndpointsMap;
+        }
+
+        @Override
+        public Endpoint selectNow(ClientRequestContext ctx) {
+            return endpointSelector.selectEndpoint();
+        }
+
+        @VisibleForTesting
+        WeightedRandomDistributionEndpointSelector endpointSelector() {
+            return endpointSelector;
+        }
+
+        private void updateEndpoints(List<Endpoint> newEndpoints) {
+            // Use the executor so the order of endpoints change is guaranteed.
+            executor.execute(() -> updateEndpoints0(newEndpoints));
+        }
+
+        // Only executed by the executor.
+        private void updateEndpoints0(List<Endpoint> newEndpoints) {
+            unhandledNewEndpoints = null;
+            if (rampingUpTaskWindowNanos > 0) {
+                // Check whether we can ramp up with the previous ramped up endpoints which are at the last
+                // of the rampingUpEndpointsEntries.
+                if (canAddToPrevEntry()) {
+                    final Set<EndpointAndStep> newlyAddedEndpoints = filterOldEndpoints(newEndpoints);
+                    if (!newlyAddedEndpoints.isEmpty()) {
+                        updateWeightAndStep(newlyAddedEndpoints);
+                        rampingUpEndpointsEntries.getLast().addEndpoints(newlyAddedEndpoints);
+                    }
+                    buildEndpointSelector();
+                    return;
+                }
+
+                // Check whether we can ramp up with the next scheduled rampingUpEndpointsEntry.
+                if (canAddToNextEntry()) {
+                    // unhandledNewEndpoints will be ramped up when updateWeightAndStep() is executed.
+                    unhandledNewEndpoints = newEndpoints;
```

Comment:
What happens if `updateEndpoints0()` is invoked before `updateWeightAndStep()` is invoked? Will the `unhandledNewEndpoints` here be set with `null` and lost?

Reply:
Yeah, it's lost and gone.

Let's say that we are ramping:
`a.com` and `b.com`
and they will be ramped up again like 100 milliseconds later.
And let's say that `updateEndpoints()` is called with `a.com`, `b.com` and `c.com`.
the new endpoints are now assigned to the `unhandledNewEndpoints`.

If `updateEndpoints()` is called again (with `a.com`, `b.com` and `d.com`) before 100 milliseconds passed, then the previous `unhandledNewEndpoints` will be gone. 
(In this case the client never send the request to `c.com`)
After 100 milliseconds passed `a.com`, `b.com` and `d.com` will be ramped up.

Reply:
OK. That makes sense.